### PR TITLE
Integrated complex_nn

### PR DESCRIPTION
**Pull request by john-m24**
Had to restructure the framework a bit to allow complex_nn to work properly but should still run other algos fine.

### CODE REVIEW

File: src/algos/complex_nn.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        # Shuffle batch elements of a to create b
+        with torch.no_grad():
+            indices = np.random.permutation(a.size(0))
+            b = a[indices]
+        
+        z, self.theta = self.real_to_complex(a,b) 
+        
+        # Get discriminator score expectation over k rotations
+        self.score_fake = 0
+        for k in self.k:
```
Comment (john-m24) [inline src/algos/complex_nn.py]: make this range(self.k)... my bad

File: src/scheduler.py [hunk 2]
```diff
@@ -66,7 +68,8 @@ def defense_test(self) -> None:
         for _, sample in enumerate(self.dataloader.test):
             items = self.utils.get_data(sample)
             z = self.algo.forward(items)
-            self.model.processing(z, items["pred_lbls"])
+            data = self.model.forward(z)
+            self.algo.infer(data,items["pred_lbls"])
```
Comment (tremblerz) [inline src/scheduler.py:72]: Can you make sure the `generate_challenge` is working too?

File: src/models/model_zoo.py [hunk 3]
```diff
@@ -45,34 +47,62 @@ def assign_model(self, config):
...
+            elif config["model_name"] == "resnet20complex":
+                self.model = ResNetProcessorComplex(3,'alpha')
         self.model = self.utils.model_on_gpus(self.model)
         self.utils.register_model("server_model", self.model)
 
     def assign_optim(self, config):
         lr = config["lr"]
-        if config["optimizer"] == "adam":
-            self.optim = torch.optim.Adam(self.model.parameters(), lr)
+        if config["model_name"] == "resnet20complex":
```
Comment (tremblerz) [inline src/models/model_zoo.py]: Why do we need an if here? I think the code is same for creating optimizers

File: src/algos/simba_algo.py [hunk 4]
```diff
@@ -50,6 +50,9 @@ def init_optim(self, config, model):
     def put_on_gpus(self):
         self.client_model = self.utils.model_on_gpus(self.client_model)
 
+    def infer(self,data,labels):
```
Comment (tremblerz) [inline src/algos/simba_algo.py:53]: great!

File: src/algos/complex_nn.py (hunk 1 above)
Comment (tremblerz) [reply src/algos/complex_nn.py]: Can you do it?

**Review by tremblerz — CHANGES_REQUESTED** [review]
